### DISCUSSION TITLE
Using Apple M1 GPU error

### DISCUSSION THREAD

**Discussion by marcoseraphin** [mrdbourke/pytorch-deep-learning#148] [General]
Hi Daniel,

I just started the course and already set up PyTorch locally on my MacBook Pro M1 machine using VS Code.
It seems to run without errors.

But if I try to move a tensor to the Apple GPU (MPS) I am getting an error using your code:

```python
print(f"PyTorch version: {torch.__version__}")

# Check PyTorch has access to MPS (Metal Performance Shader, Apple's GPU architecture)
print(f"Is MPS (Metal Performance Shader) built? {torch.backends.mps.is_built()}")
print(f"Is MPS available? {torch.backends.mps.is_available()}")

# Set the device      
device = "mps" if torch.backends.mps.is_available() else "cpu"
print(f"Using device: {device}")

# Create data and send it to the device
x = torch.rand(size=(3, 4)).to(device)
print(x)
```

==> output / error:

PyTorch version: 1.13.0.dev20220827
Is MPS (Metal Performance Shader) built? True
Is MPS available? True
Using device: mps
/Users/marcoseraphin/opt/miniconda3/envs/torch/lib/python3.9/site-packages/torch/_tensor_str.py:114: UserWarning: The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1661584288993/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(
tensor([[0.4496, 0.6186, 0.7236, 0.4368],
        [0.7587, 0.4787, 0.1247, 0.4760],
        [0.4457, 0.3475, 0.8220, 0.5320]], device='mps:0')

How do I solve the error that the tensor x is NOT transferred to the GPU ?

Thanx

Marco

**Comment by FreeBlues** [comment]
It is not an error, it is warning, tell you the `aten::masked_select` can only run on CPU.
In fact your code is now running on GPU except a bit(`aten::masked_select` etc. )

**Comment by mrdbourke** [comment]
Hi @marcoseraphin,

I get this error too.

PyTorch on MPS is still a work in progress.

However, the latest version of PyTorch (released October 2022), has added some more support for MPS devices - https://pytorch.org/blog/new-library-updates-in-pytorch-1.13/

**Reply by marcoseraphin** [reply]
Hi Daniel,

thanx...yes I already updated to 1.13 on my Mac. Looks much better, I do not get any warnings or errors since then.
I am currently working thru your great PyTorch course on ZTM.

Really impressive. I did my master thesis back in 1993 about neural networks (on a 80386 PC with MS Dos, Windows 1.0 was just released but not much used) and tried to be in touch with ML since then.

Best

Marco
